### DISCUSSION TITLE
Modbus API

### DISCUSSION THREAD

**Discussion by andrewmarles** [grblHAL/core#323] [Show and tell]
Just curious about any details (planned usage etc) on the new Modbus API.  Out of interest I have been working on a Modbus RS485 peripheral board that uses the existing VFD modbus driver and allows for events and other functions to be sent to the PicoHAL board so that behaviours (such as fancy LED effects, peristaltic pumps, fans etc) can be programmed using MicroPython which is much more accessible for many people and also offloads such things from complicating the GRBLHAL core.

I also wanted to work on adding driver support to use the pins on the PicoHAL as aux outputs/inputs (not IRQ capable) so that they can be used for simple functions, again via R485.

Plugin source code:
https://github.com/Expatria-Technologies/Plugins_picoHAL

Micropython firmware:
https://github.com/Expatria-Technologies/picohal-firmware

Board is close to release and will be fully open source like the rest of our designs.

**Comment by terjeio** [comment]
It is intended for plugin code that wants to use Modbus to communicate with external devices, just like before, but now with the option of adding Modbus TCP as a transport transparently.

Your plugin (and the spindle plugin) is a good example of what it can be used for. Even more advanced plugins (yours later on?) could have settings to allow configurable Modbus messages much like modvfd? This would open access to off-the-shelf Modbus devices both networked (TCP) and RS485 (RTU) based.

I am contemplating adding a TCP server implementation later that would allow reading/writing from/to the controller, again for plugin code to use. This could be complemented with a Modbus RTU device (slave) implementation, both wrapped in the new API. However, this would require the board to have a free serial port so would be of limited use?

> Board is close to release and will be fully open source like the rest of our designs.

:+1:
